src/encoding.rs

```diff
@@ -1,20 +1,17 @@
-use std::convert::{TryInto};
-
-pub type Padding = u32;
-pub const PADDING: usize = std::mem::size_of::<Padding>();
+use integer_encoding::VarInt;
 
 /// Encode a message, returning the bytes that must be sent before the message.
-pub fn encode_size(message: &[u8]) -> [u8; PADDING] {
-    (message.len() as Padding).to_le_bytes()
+/// A buffer is used to avoid heap allocation.
+/// It is size 10=ceil(64/7), the max required bytes to encode a u64.
+pub fn encode_size<'a>(message: &[u8], buf: &'a mut [u8; 10]) -> &'a [u8] {
+    let varint_size = message.len().encode_var(buf);
+    &buf[..varint_size]
 }
 
 /// Decodes an encoded value in a buffer.
-/// The function returns the message size or none if the buffer is less than [`PADDING`].
-pub fn decode_size(data: &[u8]) -> Option<usize> {
-    if data.len() < PADDING {
-        return None
-    }
-    data[..PADDING].try_into().map(|encoded| Padding::from_le_bytes(encoded) as usize).ok()
+/// The function returns the message size or none if the buffer is too small.
+pub fn decode_size(data: &[u8]) -> Option<(usize, usize)> {
+    usize::decode_var(data)
```

Comment:
Can it be an `u64` to make it compatible with x32 machines?

Reply:
mmm... You're right. At this point, there are two options. Force to be always u32, to encourage the compatibility, or use `usize` to allow 64bit machines to send longer messages. I prefer the safer way, and left it as `u32` because send this kind of long messages is not really recommended because if the connection is lost you need to send again all that information. So, is better to split in several messages even in 64bit machines.

What do you think?

Reply:
As you said, the data should probably be fragmented by hand anyway so it doesn't really matter in the end. The code with usize everywhere _is_ slightly simpler though because there's no casts.

Reply:
It is ok both options to me 😃